raiden/storage/restore.py

```diff
@@ -210,3 +220,87 @@ def balance_proof_query_from_keys(prefix: str, filters: Dict[str, Any]) -> Dict[
     for key, value in filters.items():
         transformed_filters[f"{prefix}balance_proof.{key}"] = value
     return transformed_filters
+
+
+def try_to_match_locksroots(
+    storage: SerializedSQLiteStorage,
+    latest_channel_state: NettingChannelState,
+    try_our_locksroot: Locksroot,
+    try_partner_locksroot: Locksroot,
+) -> Optional[LocksrootRecord]:
+    """Search in the database for `try_our_locksroot` and
+    `try_partner_locksroot`, if both are match the expected end of the channel
+    `latest_channel_state` then LocksrootRecord is returned.
+
+    Notes:
+        - The empty hash has to be handled. Because this value is the default,
+          it is always assumed to be valid.
+        - The locksroot does not have to match the latest state of the channel,
+          however it must have been part of some version of the channel. This
+          is necessary to handle cases were a malicious node closes the channel
+          with older balance proofs, otherwise settling would not be possible.
+    """
+    our_locksroot = None
+    partner_locksroot = None
+
+    if try_our_locksroot in (EMPTY_HASH, LOCKSROOT_OF_NO_LOCKS):
+        our_locksroot = LOCKSROOT_OF_NO_LOCKS
+    else:
+        event_record = get_event_with_balance_proof_by_locksroot(
+            storage=storage,
+            canonical_identifier=latest_channel_state.canonical_identifier,
+            locksroot=try_our_locksroot,
+            recipient=latest_channel_state.partner_state.address,
+        )
+
+        if event_record is not None:
+            our_locksroot = try_our_locksroot
+
+    if try_our_locksroot in (EMPTY_HASH, LOCKSROOT_OF_NO_LOCKS):
+        partner_locksroot = LOCKSROOT_OF_NO_LOCKS
+    else:
+        state_change_record = get_state_change_with_balance_proof_by_locksroot(
+            storage=storage,
+            canonical_identifier=latest_channel_state.canonical_identifier,
+            locksroot=try_partner_locksroot,
+            sender=latest_channel_state.partner_state.address,
+        )
+
+        if state_change_record is not None:
+            partner_locksroot = try_partner_locksroot
+
+    if our_locksroot is not None and partner_locksroot is not None:
+        return LocksrootRecord(our_locksroot=our_locksroot, partner_locksroot=partner_locksroot)
+
+    return None
+
+
+def order_locksroot(
+    storage: SerializedSQLiteStorage,
+    latest_channel_state: NettingChannelState,
+    locksroot1: Locksroot,
+    locksroot2: Locksroot,
+) -> Optional[LocksrootRecord]:
+    """Finds order of locksroot1 and locksroot2.
+
+    Returns None if any of the two locksroots is unknown, otherwise return
+    LocksrootRecord with the end of the locksroot correctly set.
```

Comment:
I don't understand "the end of the locksroot".  The purpose of the locksroot, or the deadline of the locksroot, or something like that?

Reply:
In the state machine logic, the channel state has these fields:

https://github.com/raiden-network/raiden/blob/fa259137bb6f21810d69e6075d79c0ca0b50403d/raiden/transfer/state.py#L344-L345

That are instances of `NettingChannelEndState`. So, the idea here was to make a reference to that type, where the local running node is one end of the channel, and the remote partner node is another end. Do you have suggestions on how this sentence can be clarified?

Reply:
"the channelend of the locksroot"